### PR TITLE
chore: adding test for multiple AddAllocation's

### PR DESCRIPTION
**Pull request by carolinaecalderon**
## Description
As a preliminary step to fixing the allocation bug where allocations with `end_time < start_time` are found & stored, write a unit test to recreate the buggy behavior. The hypothesis is that for allocations that have already ended, if master shuts down & restarts before the allocation is marked as started, the allocation will re-start & update its start time, but not reset its end time (as its set from when the master shuts down).

## Test Plan
Pass the included test (once it's merged).

## Commentary (optional)
A separate PR fixes the bug: https://github.com/determined-ai/determined/pull/7928
For the time being, this PR compares merging into `main`, which doesn't have the fix yet, but will prove that the test fails under the current `AddAllocation` implementation.

As for the order of merging commits;
- first, #7928  with changes ONLY to postgres_tasks.go.
- second, rebase this branch against main to pick up the changes. This test will now pass.
- finally, merge changes ONLY to postgres_tasks_intg_test.go.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
DET-9496

### CODE REVIEW

File: master/internal/db/postgres_tasks_intg_test.go [hunk 1]
```diff
@@ -25,6 +26,57 @@ import (
...
+
+	// Check that it's in the table.
+	_, err = db.AllocationByID(mockAllocation.AllocationID)
+	require.NoError(t, err)
+
+	// Update the allocation's end time manually.
+	_, err = Bun().NewUpdate().Table("allocations").
+		Where("allocation_id = ?", mockAllocation.AllocationID).
+		Set("end_time = ?", time.Now().UTC()).Exec(context.TODO())
+	require.NoError(t, err)
```
Comment (corban-beaird) [inline master/internal/db/postgres_tasks_intg_test.go:59]: nit: IIRC we have functions that will set the allocation's end_time ([CompleteAllocation](https://github.com/determined-ai/determined/blob/7c765ae9afa0228511c50081eeac2aefc4d13628/master/internal/db/postgres_tasks.go#L140)). Is there a reason you went with manually setting setting the end_time?

**Review by corban-beaird — APPROVED** [review]
Looks great! I only had one question, but it was a nit

File: master/internal/db/postgres_tasks_intg_test.go (hunk 1 above)
Comment (carolinaecalderon) [reply master/internal/db/postgres_tasks_intg_test.go:59]: good catch

File: master/internal/db/postgres_tasks_intg_test.go [hunk 2]
```diff
@@ -25,6 +25,56 @@ import (
...
+
+	// Update the allocation's end time.
+	mockAllocation.EndTime = ptrs.Ptr(time.Now().UTC())
+	err = db.CompleteAllocation(&mockAllocation)
+	require.NoError(t, err)
+
+	// Now check that the end time is greater than the start time.
+	a, err := db.AllocationByID(mockAllocation.AllocationID)
+	require.NoError(t, err)
+	require.Greater(t, fmt.Sprint(a.EndTime), fmt.Sprint(a.StartTime))
```
Comment (erikwilson) [inline master/internal/db/postgres_tasks_intg_test.go:62]: might want to use something like `a.EndTime.After(a.StartTime)`

File: master/internal/db/postgres_tasks_intg_test.go [hunk 3]
```diff
@@ -25,6 +25,56 @@ import (
...
+	// when a cluster crashes/master restarts, and the allocation starts over
+	// again, even if the original allocation already completed.
+	mockAllocation.StartTime = ptrs.Ptr(time.Now().UTC())
+	err = db.AddAllocation(&mockAllocation)
+	require.NoError(t, err)
+
+	a, err = db.AllocationByID(mockAllocation.AllocationID)
+	require.NoError(t, err)
+	require.Nil(t, a.EndTime)
+	require.NotNil(t, a.StartTime)
```
Comment (erikwilson) [inline master/internal/db/postgres_tasks_intg_test.go:75]: may also want to check which specific start time is being used, having static times could also help there.

File: master/internal/db/postgres_tasks_intg_test.go (hunk 2 above)
Comment (stoksc) [reply master/internal/db/postgres_tasks_intg_test.go:62]: happens to be covered by https://github.com/uber-go/guide/blob/master/style.md#use-time-to-handle-time

**Pull request closed without merging**